drivers/can/can_sam0.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2022 Vestas Wind Systems A/S
+ * Copyright (c) 2021 Alexander Wachter
+ * Copyright (c) 2022 Kamil Serwus
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+
+#include "can_mcan.h"
+
+LOG_MODULE_REGISTER(can_sam0, CONFIG_CAN_LOG_LEVEL);
+
+#define DT_DRV_COMPAT atmel_sam0_can
+
+struct can_sam0_config {
+	void (*config_irq)(void);
+	const struct pinctrl_dev_config *pcfg;
+	volatile uint32_t *mclk;
+	uint32_t mclk_mask;
+	uint16_t gclk_core_id;
+	int divider;
+};
+
+struct can_sam0_data {
+	struct can_mcan_msg_sram msg_ram;
+};
+
+void can_sam0_line_x_isr(const struct device *dev)
+{
+	can_mcan_line_0_isr(dev);
+	can_mcan_line_1_isr(dev);
+}
+
+static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_cfg = dev->config;
+	const struct can_sam0_config *sam_cfg = mcan_cfg->custom;
+
+	*rate = SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / (sam_cfg->divider);
+
+	return 0;
+}
+
+static void can_sam0_clock_enable(const struct can_sam0_config *cfg)
+{
+	/* Enable the GLCK1 with DIV*/
+	GCLK->GENCTRL[1].reg = GCLK_GENCTRL_SRC(GCLK_GENCTRL_SRC_OSC48M)
+			     | GCLK_GENCTRL_DIV(cfg->divider)
+			     | GCLK_GENCTRL_GENEN;
```

Comment:
All other SAM parts put this level of clock configuration in the SOC's code, e.g: [`.../soc_samc2x.c`](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samc2x.c)

Given that `cfg->divider` comes from each channel in the device tree, I'm also concerned that in applications that use both channels, there could be a silent incompatibility between the two... the device tree currently suggests that these dividers are independent, when they aren't!

I note that the f<sub>GCLK_CANn</sub> is listed as "_48 MHz max_", and the prescaler in the CAN peripheral is fairly limited, so is probably required...

I'm concerned that placing the GCLK1 configuration in the driver will make it less visible and potentially introduce issues in the future...
- At minimum, please add a comment to `soc_samc2x.c` to indicate that GCLK1 is reserved for CAN
- My preference would be to move configuration of GCLK1 into `soc_samc2x.c`
- Consider using two GCLK channels (one for each CAN interface), or perhaps move the `divider` value up into the `can` / `bosch,m_can-base` node

e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/ed2c25effccc55e459e2b96f389d9b1a177daf7d/soc/arm/atmel_sam0/common/soc_saml2x.c#L26-L29

Reply:
Ah, well found... Using `GCLK0` @ 48 MHz sounds sensible to me...

Reply:
@attie-argentum  @nandojve 
About clocks I inspired with can_sam and can_mcan.
In my project I need divider to set valid ntseg (i had to set specific ntseg1/ntseg2 values for specific clock speed).

Can I change can0 to config GLCK7 instead of GLCK1? It will be ok for you all?

Reply:
You mean reserve the GCLK[7] for CAN? I don't see problems with that but you need make sure all SAM0 series enable GCLK[7] at soc. Is that what you want to do, right?

Reply:
@nandojve as you can see I "dynamicly" enable at cansam0 driver. So that it isn't need to to enable at soc.
BTW In my knowlage CAN module for sam0 has only samC21 series.

Reply:
Ah ok! But as a remind that ATSAME5x have CAN-FD too.
Fine for me if you want move from GCLK1 to GCLK7.

Reply:
I believe that "conversation" can be resolved :-)
Thanks for info about E5x, I didn't know.